External trigger "plugin" functions.

hjoliver opened the pull request:
Close #2339. Close #1364 

__Trigger tasks off of *anything* by having the suite daemon call arbitrary "xtrigger functions"__ that return True/False and some result - e.g. path to a new external dataset - to broadcast to the ~~cycle point~~ dependent task.

This should **replace** in-task-proxy **clock-triggers**, client/server **external triggers**, and **polling tasks**.

* xtrigger functions are called in the process pool with asynchronous callback (except for clock xtriggers which are guaranteed to be quick)
* a function won't be called again before the previous call comes back
* once an xtrigger is satisfied, the result is remembered and the function not called again
* ~~if a task has a clock xtrigger, don't bother calling other xtriggers until the clock time is up~~ [not compatible with use of clock triggers in conditional expressions]
* xtrigger functions are tracked by unique "call signature" (function name and args).  If multiple tasks depend on the same xtrigger, it gets shared among them
* function call frequency is configurable, default PT10S
* variables like suite name, run directory, task cycle point, etc. can be substituted into function arg values via string templates (example below)
* to extend functionality (trigger off of other external events) simply put a new function in `/lib/cylc/xtriggers/` (or anywhere in `sys.path`) - and use it in the suite definition immediately

hjoliver commented:
I've implemented some xtrigger functions under `lib/cylc/xtriggers/`, with example suites under `dev/suites/xtrigger/`:
 * `wall_clock(...)`  - can replace the existing clock triggers (which are internal to task proxies)
 * `suite_state(...)` - can replace suite-state polling tasks
 * `file_exists(...)` - trigger if a specific file exists
 * `match_file(...)` - trigger if a file is found to match some pattern

hjoliver commented:
Example suite with a family of tasks that all depend on a clock trigger and a suite-state trigger:
```
[cylc]
    cycle point format = %Y
[scheduling]
    initial cycle point = 2020
    [[xtriggers]]
         suite_up = suite_state(suite=upstream, task=foo, point=%(cycle_point)s):PT10S
         clock_0 = wall_clock(offset=PT0H)
   [[dependencies]]
        [[[P1Y]]]
           graph = "@clock_0 & @suite_up => FAM"
[runtime]
    [[FAM]]
    [[f_1, f_2, f_3, f_4, f_5]]
          inherit = FAM
```
![foo](https://user-images.githubusercontent.com/2362137/30240107-453c29c2-95be-11e7-9eae-898516c8f173.png)

The suite configuration is a direct representation of the function call, e.g.: 
```
def suite_state(suite, task, point, offset=None, status='succeeded',
                         message=None, cylc_run_dir=None)
```

The optional `:INTERVAL` in `label = func(args):INTERVAL` determines how often the suite daemon should call the function.

hjoliver commented:
(failing 15 tests on Travis CI - not too bad for a change of this size)

hjoliver commented:
@cylc/core - I've haven't added tests or documentation yet, pending discussion and agreement.  I am fairly convinced that this will be super-useful though...

hjoliver commented:
A real-world use case that we need here (in addition to the Kafka case mentioned in #2339): suites to process incoming satellite data currently require either polling tasks that wait for the new datasets, or an external daemon that does that and calls `cylc ext-trigger` for every new dataset.  We'll be able to do this now with a simple xtrigger in the graph.

hjoliver commented:
Satisfied xtriggers are now stored and reloaded from the DB on restart, so the corresponding functions won't be called again (for the same function signature).  This is needed because some xtriggers will only report in satisfied once. E.g. for triggering off of a message in a broker like RabbitMQ (but not Kafka) in which a message disappears once consumed; and for triggering of detection of a new external dataset that does not have cycle point in the filename (e.g. satellite datasets) so the xtrigger function probably needs to rename the file on detection to ensure that it doesn't get re-detected in the next suite cycle.

oliver-sanders commented:
Looks good! Having external triggers appear in suite visualisations should make things much clearer.

This opens up using clock triggers with conditional statements which could be interesting e.g:

```
early | ( @clock_0 & data ) => process
```

In combination with #2334:
```
( ensemble:succeed-90% & @clock_0 ) | ( ensemble:succeed-80% & @clock_1 ) => process
```

Would it be possible to move a little more configuration into the graph by specifying selected arguments there? E.G:

```
[[[PT3H]]]
    graph = @clock => foo
[[[PT6H]]]
    graph = foo | @clock(+PT3H) => bar

hjoliver commented:
@oliver-sanders - good suggestions, I'll address them as soon as possible.  Unfortunately these new triggers can't be used in conditional expressions yet, but I hope it is easy to do - it may need #2413

hjoliver commented:
@oliver-sanders - after a bit of thought I'd prefer to leave your suggestions to possible future enhancements.

 * __clock triggers in conditional expressions__: this is obviously desirable, but I think it requires unifying the various trigger types.  This PR is still a significant improvement without that (and validation catches any attempt to do it)
 *  __allowing some xtrigger function args in the graph__: this may be desirable as a small efficiency enhancement in suites with lots of xtriggers, as you'd only have to define the function call once. However, it doesn't really add more info to the graph if you use labels carefully (e.g. `clock_PT0H => foo` vs `clock(PT0H) => foo`) and on the downside it breaks the useful one-to-one mapping between labels and unique function signatures, which is used this to qualify functional results in case a single task has multiple xtriggers of the same type, e.g.:
```
x1 = xtrig(file=a)
x2 = xtrig(file=b)
...
graph = "x1 & x2 => foo"
```
if the function `xtrig(file=a)` returns `{name: 'bob'}` and `xtrig(file=b)` returns `{name: 'alice'}` then the suite daemon will broadcast to task `foo` environment, `x1_name=bob` and `x2_name=alice`.  If instead we had this:
```
x = xtrig()
...
graph = "x(file=a) & x(file=b) => foo"
```
then it's not clear how to pass function results on to triggered tasks.

hjoliver commented:
Revised example suite:
```
[cylc]
    cycle point format = %Y
[scheduling]
    initial cycle point = 2020
    [[xtriggers]]
         suite_up = suite_state(suite=upstream, task=foo, point=%(task_cycle_point)s):PT10S
         clock_0 = wall_clock(offset=PT0H)
   [[dependencies]]
        [[[P1Y]]]
           graph = "@clock_0 & @suite_up => FAM"
[runtime]
    [[FAM]]
    [[f_1, f_2, f_3, f_4, f_5]]
          inherit = FAM
```
![bar](https://user-images.githubusercontent.com/2362137/31212516-6eb2eeaa-a9fd-11e7-856a-5bc488160737.png)

hjoliver commented:
An interesting quirk to note: **unlike tasks, xtriggers do not have an associated cycle point** (hence they are not labeled with cycle point in the graph - see image above - but for convenience they are plotted per cycle point).  Whether or not they are cycle-point specific depends on whether or not the cycle point of the waiting task is used in constructing their function args (see the template `%(task_cycle_point)s` in the example suite above).  The daemon groups xtriggers by unique function call signature, so if the cycle point does not appear in the argument list, then a single call will satisfy all cycle points as the suite runs.  Similarly, if task name appears in the argument list (via template) xtrigger calls will be task-name specific (i.e. separate calls will be made to satisfy tasks with different names).

hjoliver commented:
I will add some tests, then request a code review...

hjoliver commented:
A note on the trigger function interface. The function signature is arbitrary, except that only keyword args are allowed.  The suite-specific use of a trigger function is defined by the suite.rc syntax.  Some suite and task variables (such as cycle point) are made available for use in the function args (in the suite.rc) - these are what makes a trigger cycle-point, or task-name, or whatever, specific.  To illustrate, use of task name in the arg list below makes family members trigger at different random times off of (seemingly) the same xtrigger:

```
[scheduling]
   [[xtriggers]]
        rnd_by_name = xrandom(percent=50, secs=2, _=%(task_name)s)
   [[dependencies]]
        graph = "@rnd_by_name => FAM" 
[runtime]
   [[FAM]]
   [[f_1, f_2, f_3, f_4, f_5]]
         inherit = FAM
```

hjoliver commented:
Branch rebased onto master.

hjoliver commented:
@matthewrmshin @oliver-sanders - I think this is ready for review.  I have not added user guide documentation yet in case review results in significant changes (although I'm pretty confident that the suite.rc interface is good) ... so let me know if the functionality and usage is not clear after reading the above comments, the docstring in `lib/cylc/xtrigger_mgr.py`, functions in `lib/cylc/xtriggers/` and suites in `/dev/suites/xtrigger/` and `tests/xtriggers/`.

matthewrmshin commented:
We can ignore Codacy's failure here. The SQL statement is safe in the context it is used - adding a class level constant to the statement. The use of pickle is also safe in the context it is used. (I wonder if we should switch to using JSON instead of pickle for (de-)serializing data? I don't think Codacy complains against JSON.)

hjoliver commented:
Re codacy: yes, I guess JSoN is not a security risk like pickle, although slower (where that matters). 

BTW I still intend to go through *all* issues flagged by codacy and "ignore" or address them (particularly security ones) ... but have not found the time for that yet.

matthewrmshin commented:
Sorry, I'd try to give this one a bit more priority next week.

An initial comment with the style, however. When I did the refactor to create all the various manager classes, there was some consideration to ensure that we have a good hierarchy - so we don't end up passing too many objects around. It may not be straightforward introducing a new manager object such as in this change, but I think we should still be able to give it a bit of a tidy, so we don't end up introducing unused or used-once attributes and local variables.

hjoliver commented:
As per an email conversation offline, I will consider whether or not we can refactor a bit to avoid (for example) having to instantiate several manager classes for rather trivial reasons in the `cylc submit` command - which might suggest some kind of design deficiency.

oliver-sanders commented:
Profiling results as at 9691581:

**complex**

| Version                 | Run            | Elapsed Time (s)  | CPU Time - Total (s)  | Max Memory (kb) |
| --- | --- | --- | --- | --- |
| hjoliver/ext-trig-plug  | complex suite  | 1132.5            | 313.3                 | 67340.0         |
|master                  | complex suite  | 1031.2            | 306.5                 | 65968.0   |

**complex-validate**

| Version                 | Run            | Elapsed Time (s)  | CPU Time - Total (s)  | Max Memory (kb) |
| --- | --- | --- | --- | --- |
| hjoliver/ext-trig-plug  | complex suite  | 2.3               | 2.2                   | 32676.0         |
| master                  | complex suite  | 2.2               | 2.1                   | 32028.0    |

**diamond**

![cpu-time](https://user-images.githubusercontent.com/16705946/32842799-f01c9706-ca15-11e7-9d85-a3c72e842dbf.png)
![memory](https://user-images.githubusercontent.com/16705946/32842800-f0949512-ca15-11e7-831b-452c42362cc5.png)

hjoliver commented:
@oliver-sanders - thanks for doing the profiling experiments.   Does your comment "no noticeable difference" refer just to "clock trigger vs xtrigger" or to the diamond and complex suite results too?

The differences don't look very significant.  I guess I would expect a little more memory use because of the new module to import; very little impact on performance for suites that do not use xtriggers. 

Suites that use **a lot** of same-time clock triggers should be more efficient under the new implementation, because the new ones are shared (one call for all dependent tasks) vs the old ones are duplicated across all dependent tasks).  However, you have to use the clock triggers in the right way, so that the suite can identify the commonality (identical function arg list for all shared xtriggers).

oliver-sanders commented:
@hjoliver The "no noticeable difference" refers to testing the same example on your branch using the wall_clock xtrigger or the clock trigger mechanism.

Because the increase in memory usage scales with number of tasks (see results for diamond) i would expect that it is due to the extra variables added to the task state and taskdef classes.

matthewrmshin commented on tests/xtriggers/00-wall_clock.t:
```diff
@@ -0,0 +1,58 @@
+. $(dirname $0)/test_header
+
+run_suite() {
+  cylc run --debug $1 -s START=$2 -s HOUR=$3 -s OFFSET=$4
```
Need `--no-detach` these days.

hjoliver replied:
Oops, I guess I forgot to check test battery results and clean up after last rebase to master! (it was all passing before that - sans the Kafka problem above).

hjoliver replied:
done

matthewrmshin commented on dev/suites/xtrigger/kafka/consumer/suite.rc:
```diff
@@ -0,0 +1,52 @@
+#------------------------------------------------------------------------------
+# A suite that uses the cylc_kafka_consumer module as an external trigger
+# function, to trigger tasks off of Kafka messages. Jinja2 inputs are defined
+# in rose-suite.conf.
```
This suite is causing `./tests/validate/04-builtin-suites.t` to fail if `kafka` is not installed.

hjoliver replied:
Sorry, I meant to move this to another location. It should probably be an example of how to trigger off of Kafka messages, rather than an officially supported interface.

matthewrmshin commented on tests/xtriggers/02-persistence.t:
```diff
@@ -0,0 +1,62 @@
+
+# Restart the suite, to run the final cycle point.
+TEST_NAME=${TEST_NAME_BASE}-restart
+suite_run_ok ${TEST_NAME} cylc restart --no-detach $SUITE_NAME
```
This test is failing in my environment with this traceback in the suite's err file:

```
2017-11-17T11:06:02Z CRITICAL - Traceback (most recent call last):
	  File "/home/matt/cylc.git/lib/cylc/scheduler.py", line 235, in start
	    self.configure()
	  File "/home/matt/cylc.git/lib/cylc/scheduler.py", line 382, in configure
	    self.load_tasks_for_restart()
	  File "/home/matt/cylc.git/lib/cylc/scheduler.py", line 460, in load_tasks_for_restart
	    self.xtrigger_mgr.load_xtrigger_for_restart)
	  File "/home/matt/cylc.git/lib/cylc/rundb.py", line 639, in select_xtriggers_for_restart
	    callback(row_idx, list(row))
	  File "/home/matt/cylc.git/lib/cylc/xtrigger_mgr.py", line 123, in load_xtrigger_for_restart
	    self.sat_xtrig[sig] = pickle.loads(results_pickle)
	  File "/usr/lib64/python2.6/pickle.py", line 1374, in loads
	    return Unpickler(file).load()
	  File "/usr/lib64/python2.6/pickle.py", line 858, in load
	    dispatch[key](self)
	KeyError: '\x00'
```

hjoliver replied:
Huh, OK.

hjoliver replied:
(hopefully this is fine now, using json instead of pickle)

hjoliver commented on lib/cylc/mp_pool.py:
```diff
@@ -89,6 +87,26 @@ def _run_command(ctx):
+    LOG.debug(ctx)
+    kwargs = ctx.kwargs
+    try:
+        ctx.ret_val = ctx.func(**kwargs)
```
Yeah, that occurred to me too - do you know a good way to do it?

oliver-sanders replied:
Could use signal: https://docs.python.org/3/library/signal.html#example

```python
import signal
import time


def timeout_function(func, time):
    def timeout_callback(foo, bar):
        raise KeyboardInterrupt()

    signal.signal(signal.SIGALRM, timeout_callback)
    signal.alarm(time)
    try:
        func()
    except KeyboardInterrupt:
        print "Timed out!"
    finally:
        # Disable the alarm.
        signal.alarm(0)

def function():
    itt = 1
    while True:
        itt += 1
        print '#', itt
        time.sleep(0.1)

timeout_function(function, 5)
```

matthewrmshin replied:
(Off topic, but this is one of the reasons why I am keen to implement our own multi-process pool - so we can have much better control of child processes.)

hjoliver replied:
Timeout implemented via signals as suggested; works well.

hjoliver replied:
Note function timeout is no longer necessary. After the process pool refactor - no persistent pool processes anymore - I've had to execute these functions as processes via a small wrapper (i.e. turn a function into a program), and we now have timeout on pool processes.

hjoliver commented on lib/cylc/taskdef.py:
```diff
@@ -77,10 +79,24 @@ def __init__(self, name, rtcfg, run_mode, start_point, spawn_ahead):
         self.external_triggers = []
+        self.xtrig_labels = set()
+        self.xclock_label = None
+        self.xclock_offset_secs = None
```
Are would so concerned about memory footprint that it is better to repeatedly re-compute a single value?

hjoliver replied:
Done.

hjoliver commented on lib/cylc/xtrigger_mgr.py:
```diff
@@ -0,0 +1,247 @@
+        ctx = deepcopy(self.clockx_map[label])
+        ctx.kwargs.update(
+            {
+                'point': str(itask.point),
```
True.

hjoliver replied:
fixed

hjoliver commented on lib/cylc/xtrigger_mgr.py:
```diff
@@ -0,0 +1,247 @@
+        ctx.kwargs.update(
+            {
+                'point': str(itask.point),
+                'point_as_seconds': itask.get_point_as_seconds(),
```
There might be a requirement, but point is provided (via string templating) and functions can therefore compute point_as_seconds if needed ... however, I suppose it might be better to provide it routinely.

hjoliver commented:
Function timeout done.  Other feedback still to be addressed.

hjoliver commented:
(I'm back on contract for the project that motivated this, time to deal with some nasty conflicts...)

TomekTrzeciak commented:
I like the idea of having triggers as nodes in the graph, but I wonder if there is perhaps scope for a more generic approach than proposed in this PR. 

The way I think about it is that nodes in Cylc graph symbolise some kind of execution, typically tasks run as (possibly remote) jobs. Xtriggers would be another type of execution unit and action nodes from #2333 (which could replace event handles) are yet another type. 

For each type of the execution unit one might have different requirements, defaults and restrictions, but their behaviour can be broadly classified by according to:
- what is run (python function, shell script, maybe even entire sub-suite as per #2241)
- where and how the execution happens (locally, remotely, through batch system submission)
- where the output goes (stdout/stderr, managed log files)

I would argue that the above classification generally forms an orthogonal solution space and one might reasonably want to mix and match different elements from it to achieve the desired behaviour. For example one might want to execute the file polling xtrigger remotely on a host where the files are or send the output from all tasks to global stdout/stderr in small suites (or sub-suites nested as tasks in a larger suite). 

So perhaps rather than adding new configuration section and graph syntax for xtrigger nodes, the meaning of the graph nodes the could be generalised and the runtime definition extended to support these additional modes of behaviour (and the future ones, like action nodes). It would be easy to introduce a new setting like, say `mode = XTRIGGER` or `type = XTRIGGER`, or perhaps one could even achieve that through some built-in (meta) families with  something like `inherit = XTRIGGER`.

hjoliver commented:
@TomekTrzeciak - I think that may be a very good idea, but it would widen the scope of this PR a lot, and we would need to think through the many implications carefully.  I have a fairly strong requirement to get the existing functionality in soon because a major site is building infrastructure that depends on it, so - if others agree - I suggest we punt your suggestion to a new issue for future consideration.  (I do however still intend to address some of the interface suggestions raised by @oliver-sanders above).

TomekTrzeciak commented:
@hjoliver , fair enough, I wanted to put this up for discussion now only because changing from this PR to something more generic later means that there might be more migration and backwards compatibility issues to tackle.

hjoliver commented:
@TomekTrzeciak - that may be true, but I don't think it would be possible at the moment - given current resourcing and other more critical development priorities - to agree on all the details of your proposal, and then implement it, in a reasonably short time frame.

hjoliver commented on lib/cylc/xtriggers/wall_clock.py:
```diff
@@ -0,0 +1,30 @@
+from cylc.cycling.iso8601 import interval_parse
+
+
+def wall_clock(point, point_as_seconds, offset):
```
Done.

hjoliver commented:
@oliver-sanders - all minor issues addressed, I think.  On your main review points above https://github.com/cylc/cylc/pull/2423#pullrequestreview-74734086:

> A few comments on the function interface
- built-in zero-offset `wall_clock` trigger - done
- integer, float, boolean args - done 
- *Rather than using string templates we could automatically provide kwargs if referenced* - can you explain this please (email if you like) - I've not used `rose macro` and this wasn't obvious to me from a quick look at the docs.

> Possible future extensions
- some arguments in the graph - agreed (for future) - graph parsing may be difficult (parentheses are used for grouping already) and need to decided how to pass functions results to tasks when the xtrigger labels are not unique 
- conditional statements - definitely agreed (for future, prob requires unifying all trigger types) - will document prominently for now
- function signature validation - agreed (for future) - will also document this as a limitation for now

> Other points
- *The inevitable mistake* - done (trigger funcs now take both args and kwargs)
- syntax file updates - it seems not needed, from a quick look

hjoliver commented:
TODO:
 - address a small number of broken tests 
 - document

Hoping to get this done this week...

oliver-sanders commented:
> Rather than using string templates we could automatically provide kwargs if referenced

I think what I was getting at here was probably more to do with automatically providing the `debug` argument so that it doesn't have to be manually written out for each trigger (unless the user desires the value to be different to that of the suite).

hjoliver commented:
@oliver-sanders @matthewrmshin - ready for final (I hope) review, except for a few more paragraphs of documentation.  See cover email (coming...)

matthewrmshin commented on bin/cylc-help:
```diff
@@ -292,6 +292,7 @@ admin_commands['profile-battery'] = ['profile-battery']
 admin_commands['import-examples'] = ['import-examples']
 admin_commands['upgrade-run-dir'] = ['upgrade-run-dir']
 admin_commands['check-software'] = ['check-software']
+admin_commands['wrap-func'] = ['wrap-func']
```
More like `call-func` than `wrap-func`?

hjoliver replied:
Well, "call" suggests a normal function call, doesn't it?  This wraps a function into a minimal executable program, to execute in the process pool.

hjoliver replied:
Changed to `function-run` - less ambiguous, an no abbrev similarity to other commands (`cylc warranty` was a problem with `wrap-func`; and `cylc run-func` is a no-go thanks to `cylc run`).

matthewrmshin commented on bin/cylc-submit:
```diff
@@ -101,8 +103,11 @@ def main():
+    db_mgr = SuiteDatabaseManager()
+    b_mgr = BroadcastMgr(db_mgr)
+    te_mgr = TaskEventsManager(suite, pool, db_mgr, b_mgr)
+    task_job_mgr = TaskJobManager(suite, pool, db_mgr, suite_srv_mgr, te_mgr)
```
Is this change still necessary?

hjoliver replied:
What do you mean exactly by "this change"?  Oh, maybe just that the `te_mgr` and `b_bgr` variables are unnecessary as only used once?   I'll change that...

hjoliver replied:
Done.

matthewrmshin commented on bin/cylc-wrap-func:
```diff
@@ -0,0 +1,58 @@
+    func_name = sys.argv[1]
+    func_args = json.loads(sys.argv[2])
+    func_kwargs = json.loads(sys.argv[3])
+    run_func(func_name, func_args, func_kwargs)
```
It may be good to move the *json-load => call function => json-dump* logic to a location under `lib/` (in the future?). (See also protocols such as JSON-RPC.)

hjoliver replied:
Could do (pref "for the future" as I haven't time to look at JSON-RPC right now)

matthewrmshin replied:
Fair enough.

hjoliver replied:
I've taken your advice on move into `lib/cylc`.

matthewrmshin commented on lib/cylc/cfgspec/suite.py:
```diff
@@ -29,8 +29,9 @@
 
 from cylc.cfgspec.utils import (
-    coerce_interval, coerce_interval_list, DurationFloat)
+    coerce_interval, coerce_xtrig, coerce_interval_list, DurationFloat)
```
I may move this in #2609!

hjoliver replied:
Yes, I thought you might :+1:

matthewrmshin commented on lib/cylc/cfgspec/utils.py:
```diff
@@ -17,37 +17,50 @@
-    IllegalValueError
-)
+    IllegalValueError)
+from cylc.mp_pool import SuiteFuncContext
```
Awful name `SuiteFuncContext`. Probably my fault with `SuiteProcContext`!

hjoliver replied:
Agreed, and yes, just following the precedent!

matthewrmshin commented on lib/cylc/mp_pool.py:
```diff
@@ -17,16 +17,19 @@
+import time
+import json
+import traceback
+from signal import signal, alarm, SIGALRM
```
New imports still used?

hjoliver replied:
Yes, I'll remove those.  That was from the original function timeout mechanism, not needed now after your process pool refactor.

hjoliver replied:
Done.

matthewrmshin commented on lib/cylc/suite_db_mgr.py:
```diff
@@ -306,6 +309,13 @@ def put_task_event_timers(self, task_events_mgr):
                     "delay": timer.delay,
                     "timeout": timer.timeout})
 
+    def put_xtriggers(self, sat_xtrig):
```
Add docstring?

hjoliver replied:
will do...

hjoliver replied:
Done.

matthewrmshin commented on lib/cylc/task_state.py:
```diff
@@ -224,6 +233,14 @@ def satisfy_me(self, all_task_outputs):
+            xclock_satisfied = False
+        else:
+            xclock_satisfied = True
+        return xclock_satisfied and all(self.xtriggers.values())
```
```python
    return self.xclock and self.xclock[1] and all(self.xtriggers.values())
```

hjoliver replied:
ok...

hjoliver replied:
Addressed (slightly differently - "None and bool" returns None)

matthewrmshin commented on lib/cylc/xtrigger_mgr.py:
```diff
@@ -0,0 +1,297 @@
+            if sig not in self.all_xtrig:
+                rem.append(sig)
+        for r in rem:
+            del self.sat_xtrig[r]
```
May be?

```python
for sig in list(self.sat_xtrig):
    if sig not in self.all_xtrig:
        del self.sat_xtrig[sig]

# OR:
self.sat_xtrig = [sig for sig in self.sat_xtrig if sig in self.all_xtrig]
```

hjoliver replied:
ok ...

hjoliver replied:
Done.

hjoliver commented:
(I'm addressing some minor issues flagged by Codacy...)

hjoliver commented:
Done (I hope). Documentation completed (I'm sure it could be improved, but I'm pretty happy with it).

matthewrmshin commented:
Good to have this in early for the next-release cycle.

hjoliver commented on doc/src/cylc-user-guide/cug.tex:
```diff
@@ -3621,7 +3625,7 @@ \subsubsection{Trigger Types}
 \end{lstlisting}
 
-{\em The sequential declaration is deprecated} however, in favour of explicit
+{\em The sequential declaration is deprecated} however, in favor of explicit
```
Yes, I'm a conflicted mess about US-NZ/EN spelling.  I lived in the US for 10 years, and US spelling tends to be used in software, so I tend to "correct" myself in both directions randomly.  I suppose we should decided on one convention and stick to it.

hjoliver commented on bin/cylc-wrap-func:
```diff
@@ -0,0 +1,58 @@
+
+# Function return value to stdout: for compat with the command process pool.
+
+import importlib
```
I think you forgot to fetch my updates.  This command has changed name to `cylc-function-run`, and the unused import removed.

hjoliver commented on doc/src/cylc-user-guide/cug.tex:
```diff
@@ -6166,6 +6092,466 @@ \subsubsection{Where To Put Batch System Handler Modules}
+The return value of the \lstinline=suite_state= trigger function looks like
+this:
+\begin{lstlisting}
+    results = {
```
It can handle this already (if I understand your point correctly) because we split AND expressions into separate triggers.  It's just conditional OR expressions that are currently off-limits for external triggers.

hjoliver commented on doc/src/cylc-user-guide/cug.tex:
```diff
@@ -6166,6 +6092,466 @@ \subsubsection{Where To Put Batch System Handler Modules}
+
+The following string templates are available for use, if the trigger function
+needs any of this information, in function arguments in the suite configuration:
+\begin{myitemize}
```
This is the explanation I was looking for earlier. It does appear to be a better way, but it will have to be another PR at this point.  I have at least done a quick fix to provide suite debug mode by string template, like the other suite and task attributes, instead of automatically adding a debug kwarg in suite debug mode (which breaks a function that does not take a debug arg).

hjoliver commented on doc/src/cylc-user-guide/cug.tex:
```diff
@@ -6166,6 +6092,466 @@ \subsubsection{Where To Put Batch System Handler Modules}
+\section{External Event Triggers}
+\label{External Event Triggers}
+
+{\em WARNING: this is a new capability - its suite configuration interface may
```
Yes, fair enough.  I'll changed the text to say we don't recommend wholesale conversion until the interface has stabilized.

matthewrmshin commented:
(The latest change looks OK visually, but given that we have got a bit of changes here, I'd like to run the test battery in my normal work environment to make sure it does not break.)

hjoliver commented:
(@matthewrmshin - yes that's why I didn't hit "merge".  To explain, the final changes are:
- `get_func` (dynamic import by name)  is not needed for clock triggers because the function name is fixed (`wall_clock`)
- therefore `get_func` is only needed now in `lib/cylc/mp_pool.py`, so I moved it there
- also `get_func` was not working for local xtrigger import from `<suite-dir>/lib/python` so I fixed that, and have now modified a test to ensure that this is tested
- a couple more doc tweaks

matthewrmshin commented:
Many tests in the test battery are hanging in my environment. I'll try again and see what is going on.

hjoliver commented:
Damn, that's weird.

hjoliver commented:
Tests OK on my VM (no remote tests, but that shouldn't matter on this branch).

matthewrmshin commented:
Ran test battery again, and it was all OK. The first attempt probably hit some local glitches.

hjoliver commented:
The final issue flagged by Codacy is not relevant - this branch necessarily adds one more DB insert (and its construction from strings is not a problem in this context).  Travis CI tests passed (final commit didn't run it, as docs only).  MERGING...
